### PR TITLE
Fulltext optimizations for OpenDigi TEI files (+DatenProbleme 773g changes)

### PR DESCRIPTION
**Pull request by mtrojan-ub**

### CODE REVIEW

File: cpp/lib/src/SmartDownloader.cc [hunk 1]
```diff
@@ -189,56 +188,22 @@ bool DiglitSmartDownloader::downloadDocImpl(const std::string &url, const TimeLi
                                             std::string * const document, std::string * const http_header_charset,
                                             std::string * const error_message)
 {
+    std::string url_improved(url);
```
Comment (ruschein) [inline cpp/lib/src/SmartDownloader.cc]: improved_url would be a better name.

File: cpp/lib/src/MediaTypeUtil.cc [hunk 2]
```diff
@@ -114,6 +115,24 @@ std::string GetMediaType(const std::string &document, const bool auto_simplify)
 }
 
 
+std::string GetMediaType(const std::string &document, std::string * const subtype, const bool auto_simplify) {
+    std::string media_type = GetMediaType(document, auto_simplify);
```
Comment (ruschein) [inline cpp/lib/src/MediaTypeUtil.cc]: No initalisations with "=", please.  Also, I think "media_type" can be const.

File: cpp/lib/src/MediaTypeUtil.cc [hunk 3]
```diff
@@ -114,6 +115,24 @@ std::string GetMediaType(const std::string &document, const bool auto_simplify)
...
+
+    if (media_type == "text/xml") {
+        XMLParser parser(document, XMLParser::XML_STRING);
+
+        std::map<std::string, std::string> attrib_map;
+        XMLParser::XMLPart part;
+        if (parser.skipTo(XMLParser::XMLPart::OPENING_TAG, "", &part)) {
+            if (part.data_ == "TEI")
```
Comment (ruschein) [inline cpp/lib/src/MediaTypeUtil.cc]: This strategy seems a bit risky.  Can't we use a more robust way of determining the subtype?

File: cpp/lib/src/MediaTypeUtil.cc [hunk 4]
```diff
@@ -114,6 +115,24 @@ std::string GetMediaType(const std::string &document, const bool auto_simplify)
...
+                *subtype = "tei";
+        }
+    }
```
Comment (ruschein) [inline cpp/lib/src/MediaTypeUtil.cc]: Should this be "tei" or "TEI"?

File: cpp/lib/src/SmartDownloader.cc [hunk 5]
```diff
@@ -189,56 +188,22 @@ bool DiglitSmartDownloader::downloadDocImpl(const std::string &url, const TimeLi
...
+    if (RegexMatcher::Matched("/diglit/", url))
+        url_improved = StringUtil::ReplaceString("/diglit/", "/opendigi/", &url_improved);
+    if (RegexMatcher::Matched("/opendigi/", url_improved))
+        url_improved = url_improved + "/ocr";
```
Comment (ruschein) [inline cpp/lib/src/SmartDownloader.cc]: Please use +=, it's more efficient and also easier for humans to parse.

File: cpp/lib/src/MediaTypeUtil.cc (hunk 4 above)
Comment (mtrojan-ub) [reply cpp/lib/src/MediaTypeUtil.cc]: The returned media type is lowercase (e.g. "text/xml"), so the subtype should be lowercase as well.

File: cpp/lib/src/MediaTypeUtil.cc (hunk 3 above)
Comment (mtrojan-ub) [reply cpp/lib/src/MediaTypeUtil.cc]: now also checking xmlns

File: cpp/lib/src/SmartDownloader.cc (hunk 5 above)
Comment (mtrojan-ub) [reply cpp/lib/src/SmartDownloader.cc]: done

File: cpp/lib/src/SmartDownloader.cc (hunk 1 above)
Comment (mtrojan-ub) [reply cpp/lib/src/SmartDownloader.cc]: renamed

**Pull request merged**